## SALES DATA ANALYSIS

### Initializing Project


In [23]:
#Importing Necessary Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [24]:
df = pd.read_csv('../Dataset/sales_data.csv')
df.head(15) 


ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0         10107               30      95.70                2  2871.00   
1         10121               34      81.35                5  2765.90   
2         10134               41      94.74                2  3884.34   
3         10145               45      83.26                6  3746.70   
4         10159               49     100.00               14  5205.27   
5         10168               36      96.66                1  3479.76   
6         10180               29      86.13                9  2497.77   
7         10188               48     100.00                1  5512.32   
8         10201               22      98.57                2  2168.54   
9         10211               41     100.00               14  4708.44   
10        10223               37     100.00                1  3965.66   
11        10237               23     100.00                7  2333.12   
12        10251               28     100.00                2  3188.64   
13        10263               34     100.00                2  3676.76   
14        10275               45      92.83                1  4177.35   

          ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0    2/24/2003 0:00  Shipped       1         2     2003  ...   
1   05/07/2003 0:00  Shipped       2         5     2003  ...   
2   07/01/2003 0:00  Shipped       3         7     2003  ...   
3    8/25/2003 0:00  Shipped       3         8     2003  ...   
4   10/10/2003 0:00  Shipped       4        10     2003  ...   
5   10/28/2003 0:00  Shipped       4        10     2003  ...   
6   11/11/2003 0:00  Shipped       4        11     2003  ...   
7   11/18/2003 0:00  Shipped       4        11     2003  ...   
8   12/01/2003 0:00  Shipped       4        12     2003  ...   
9    1/15/2004 0:00  Shipped       1         1     2004  ...   
10   2/20/2004 0:00  Shipped       1         2     2004  ...   
11  04/05/2004 0:00  Shipped       2         4     2004  ...   
12   5/18/2004 0:00  Shipped       2         5     2004  ...   
13   6/28/2004 0:00  Shipped       2         6     2004  ...   
14   7/23/2004 0:00  Shipped       3         7     2004  ...   

                     ADDRESSLINE1  ADDRESSLINE2           CITY     STATE  \
0         897 Long Airport Avenue           NaN            NYC        NY   
1              59 rue de l'Abbaye           NaN          Reims       NaN   
2   27 rue du Colonel Pierre Avia           NaN          Paris       NaN   
3              78934 Hillside Dr.           NaN       Pasadena        CA   
4                 7734 Strong St.           NaN  San Francisco        CA   
5               9408 Furth Circle           NaN     Burlingame        CA   
6         184, chausse de Tournai           NaN          Lille       NaN   
7     Drammen 121, PR 744 Sentrum           NaN         Bergen       NaN   
8       5557 North Pendale Street           NaN  San Francisco        CA   
9               25, rue Lauriston           NaN          Paris       NaN   
10              636 St Kilda Road       Level 3      Melbourne  Victoria   
11              2678 Kingston Rd.     Suite 101            NYC        NY   
12                  7476 Moss Rd.           NaN         Newark        NJ   
13            25593 South Bay Ln.           NaN    Bridgewater        CT   
14   67, rue des Cinquante Otages           NaN         Nantes       NaN   

   POSTALCODE    COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0       10022        USA       NaN              Yu             Kwai    Small  
1       51100     France      EMEA         Henriot             Paul    Small  
2       75508     France      EMEA        Da Cunha           Daniel   Medium  
3       90003        USA       NaN           Young            Julie   Medium  
4         NaN        USA       NaN           Brown            Julie   Medium  
5       94217        USA       NaN          Hirano             Juri   Medium  
6       59000     France      EMEA      

In [25]:
# SHAPE AND COLUMN NAMES OF DATASET
df.shape, df.columns

((2823, 25),
 Index(['ORDERNUMBER', 'QUANTITYORDERED', 'PRICEEACH', 'ORDERLINENUMBER',
        'SALES', 'ORDERDATE', 'STATUS', 'QTR_ID', 'MONTH_ID', 'YEAR_ID',
        'PRODUCTLINE', 'MSRP', 'PRODUCTCODE', 'CUSTOMERNAME', 'PHONE',
        'ADDRESSLINE1', 'ADDRESSLINE2', 'CITY', 'STATE', 'POSTALCODE',
        'COUNTRY', 'TERRITORY', 'CONTACTLASTNAME', 'CONTACTFIRSTNAME',
        'DEALSIZE'],
       dtype='object'))

### DATA INSPECTION


In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDERNUMBER       2823 non-null   int64  
 1   QUANTITYORDERED   2823 non-null   int64  
 2   PRICEEACH         2823 non-null   float64
 3   ORDERLINENUMBER   2823 non-null   int64  
 4   SALES             2823 non-null   float64
 5   ORDERDATE         2823 non-null   object 
 6   STATUS            2823 non-null   object 
 7   QTR_ID            2823 non-null   int64  
 8   MONTH_ID          2823 non-null   int64  
 9   YEAR_ID           2823 non-null   int64  
 10  PRODUCTLINE       2823 non-null   object 
 11  MSRP              2823 non-null   int64  
 12  PRODUCTCODE       2823 non-null   object 
 13  CUSTOMERNAME      2823 non-null   object 
 14  PHONE             2823 non-null   object 
 15  ADDRESSLINE1      2823 non-null   object 
 16  ADDRESSLINE2      302 non-null    object 


In [27]:
df.isna().sum()

ORDERNUMBER            0
QUANTITYORDERED        0
PRICEEACH              0
ORDERLINENUMBER        0
SALES                  0
ORDERDATE              0
STATUS                 0
QTR_ID                 0
MONTH_ID               0
YEAR_ID                0
PRODUCTLINE            0
MSRP                   0
PRODUCTCODE            0
CUSTOMERNAME           0
PHONE                  0
ADDRESSLINE1           0
ADDRESSLINE2        2521
CITY                   0
STATE               1486
POSTALCODE            76
COUNTRY                0
TERRITORY           1074
CONTACTLASTNAME        0
CONTACTFIRSTNAME       0
DEALSIZE               0
dtype: int64

### OBSERVATION
- Number of Rows and Column: `2823, 25`
- Column-Wise Missing Values: [ADDRESSLINE2 : `2521`,
                    STATE: `1486`,
                    POSTAL CODE: `76`,
                    TERRITORY: `1074`]
- dtypes: `float64(2), int64(7), object(16)`

### DATA CLEANING

In [28]:
# Converting `ORDERDATE` to datetime

df['ORDERDATE'] = pd.to_datetime(df['ORDERDATE'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ORDERNUMBER       2823 non-null   int64         
 1   QUANTITYORDERED   2823 non-null   int64         
 2   PRICEEACH         2823 non-null   float64       
 3   ORDERLINENUMBER   2823 non-null   int64         
 4   SALES             2823 non-null   float64       
 5   ORDERDATE         2823 non-null   datetime64[ns]
 6   STATUS            2823 non-null   object        
 7   QTR_ID            2823 non-null   int64         
 8   MONTH_ID          2823 non-null   int64         
 9   YEAR_ID           2823 non-null   int64         
 10  PRODUCTLINE       2823 non-null   object        
 11  MSRP              2823 non-null   int64         
 12  PRODUCTCODE       2823 non-null   object        
 13  CUSTOMERNAME      2823 non-null   object        
 14  PHONE             2823 n

### VALIDATING SALES

In [29]:
df['CALCULATED_SALES'] = df['QUANTITYORDERED'] * df['PRICEEACH']
df.round()  #used `.round()` to round 1 decimal places.

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER   SALES  \
0           10107               30       96.0                2  2871.0   
1           10121               34       81.0                5  2766.0   
2           10134               41       95.0                2  3884.0   
3           10145               45       83.0                6  3747.0   
4           10159               49      100.0               14  5205.0   
...           ...              ...        ...              ...     ...   
2818        10350               20      100.0               15  2244.0   
2819        10373               29      100.0                1  3979.0   
2820        10386               43      100.0                4  5418.0   
2821        10397               34       62.0                1  2116.0   
2822        10414               47       66.0                9  3079.0   

      ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ... ADDRESSLINE2  \
0    2003-02-24   Shipped       1         2     2003  ...          NaN   
1    2003-05-07   Shipped       2         5     2003  ...          NaN   
2    2003-07-01   Shipped       3         7     2003  ...          NaN   
3    2003-08-25   Shipped       3         8     2003  ...          NaN   
4    2003-10-10   Shipped       4        10     2003  ...          NaN   
...         ...       ...     ...       ...      ...  ...          ...   
2818 2004-12-02   Shipped       4        12     2004  ...          NaN   
2819 2005-01-31   Shipped       1         1     2005  ...          NaN   
2820 2005-03-01  Resolved       1         3     2005  ...          NaN   
2821 2005-03-28   Shipped       1         3     2005  ...          NaN   
2822 2005-05-06   On Hold       2         5     2005  ...          NaN   

               CITY STATE POSTALCODE  COUNTRY TERRITORY CONTACTLASTNAME  \
0               NYC    NY      10022      USA       NaN              Yu   
1             Reims   NaN      51100   France      EMEA         Henriot   
2             Paris   NaN      75508   France      EMEA        Da Cunha   
3          Pasadena    CA      90003      USA       NaN           Young   
4     San Francisco    CA        NaN      USA       NaN           Brown   
...             ...   ...        ...      ...       ...             ...   
2818         Madrid   NaN      28034    Spain      EMEA          Freyre   
2819           Oulu   NaN      90110  Finland      EMEA       Koskitalo   
2820         Madrid   NaN      28034    Spain      EMEA          Freyre   
2821       Toulouse   NaN      31000   France      EMEA          Roulet   
2822         Boston    MA      51003      USA       NaN         Yoshido   

     CONTACTFIRSTNAME DEALSIZE CALCULATED_SALES  
0                Kwai    Small           2871.0  
1                Paul    Small           2766.0  
2              Daniel   Medium           3884.0  
3               Julie   Medium           3747.0  
4               Julie   Medium           4900.0  
...               ...      ...              ...  
2818            Diego    Small           2000.0  
2819           Pirkko   Medium           2900.0  
2820            Diego   Medium           4300.0  
2821          Annette    Small           2116.0  
2822             Juri   Medium           3079.0  

[2823 rows x 26 columns]

In [30]:
df.describe()

ORDERNUMBER  QUANTITYORDERED    PRICEEACH  ORDERLINENUMBER  \
count   2823.000000      2823.000000  2823.000000      2823.000000   
mean   10258.725115        35.092809    83.658544         6.466171   
min    10100.000000         6.000000    26.880000         1.000000   
25%    10180.000000        27.000000    68.860000         3.000000   
50%    10262.000000        35.000000    95.700000         6.000000   
75%    10333.500000        43.000000   100.000000         9.000000   
max    10425.000000        97.000000   100.000000        18.000000   
std       92.085478         9.741443    20.174277         4.225841   

              SALES                      ORDERDATE       QTR_ID     MONTH_ID  \
count   2823.000000                           2823  2823.000000  2823.000000   
mean    3553.889072  2004-05-11 00:16:49.989373056     2.717676     7.092455   
min      482.130000            2003-01-06 00:00:00     1.000000     1.000000   
25%     2203.430000            2003-11-06 12:00:00     2.000000     4.000000   
50%     3184.800000            2004-06-15 00:00:00     3.000000     8.000000   
75%     4508.000000            2004-11-17 12:00:00     4.000000    11.000000   
max    14082.800000            2005-05-31 00:00:00     4.000000    12.000000   
std     1841.865106                            NaN     1.203878     3.656633   

          YEAR_ID         MSRP  CALCULATED_SALES  
count  2823.00000  2823.000000       2823.000000  
mean   2003.81509   100.715551       2936.906408  
min    2003.00000    33.000000        482.130000  
25%    2003.00000    68.000000       2100.000000  
50%    2004.00000    99.000000       2800.000000  
75%    2004.00000   124.000000       3726.810000  
max    2005.00000   214.000000       9048.160000  
std       0.69967    40.187912       1105.644284

### OBSERVATION
- New Column `Calculated_sales` create to validate sales column
- `Sales` columns seems not reliable.
- For Example: `QUANTITYORDERED: 41 * PRICEEACH: 100 == 4100`, but in `SALES` column it was `4708.0`

### EXPLORATORY DATA ANALYSIS (EDA)
- Total sales per year
- Total sales per month

In [36]:
# MONTHLY SALE 
df_month_sales = df.groupby('MONTH_ID')
df_month_sales

In [37]:
df.groupby('MONTH_ID')['CALCULATED_SALES'].sum()

MONTH_ID
1      659582.29
2      668328.15
3      626186.73
4      560334.46
5      756812.91
6      384743.59
7      420973.34
8      552132.91
9      474900.12
10     919036.70
11    1744682.45
12     523173.14
Name: CALCULATED_SALES, dtype: float64

In [38]:
# YEARLY SALE
df_year_sales = df.groupby('YEAR_ID')
df_year_sales

In [39]:
df.groupby('YEAR_ID')['CALCULATED_SALES'].sum()

YEAR_ID
2003    2898149.94
2004    3913700.87
2005    1479035.98
Name: CALCULATED_SALES, dtype: float64

### CUSTOMER ORDER PERFORMANCE 
- Average order value (AOV) = Revenue/number of orders
- Largest single order
- Smallest non-zero order
- Identify Top 5 highest revenue orders.

In [ ]:
#AOV
df.aov = 

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
4           10159               49     100.00               14  5205.27   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

      ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ... ADDRESSLINE2  \
0    2003-02-24   Shipped       1         2     2003  ...          NaN   
1    2003-05-07   Shipped       2         5     2003  ...          NaN   
2    2003-07-01   Shipped       3         7     2003  ...          NaN   
3    2003-08-25   Shipped       3         8     2003  ...          NaN   
4    2003-10-10   Shipped       4        10     2003  ...          NaN   
...         ...       ...     ...       ...      ...  ...          ...   
2818 2004-12-02   Shipped       4        12     2004  ...          NaN   
2819 2005-01-31   Shipped       1         1     2005  ...          NaN   
2820 2005-03-01  Resolved       1         3     2005  ...          NaN   
2821 2005-03-28   Shipped       1         3     2005  ...          NaN   
2822 2005-05-06   On Hold       2         5     2005  ...          NaN   

               CITY STATE POSTALCODE  COUNTRY TERRITORY CONTACTLASTNAME  \
0               NYC    NY      10022      USA       NaN              Yu   
1             Reims   NaN      51100   France      EMEA         Henriot   
2             Paris   NaN      75508   France      EMEA        Da Cunha   
3          Pasadena    CA      90003      USA       NaN           Young   
4     San Francisco    CA        NaN      USA       NaN           Brown   
...             ...   ...        ...      ...       ...             ...   
2818         Madrid   NaN      28034    Spain      EMEA          Freyre   
2819           Oulu   NaN      90110  Finland      EMEA       Koskitalo   
2820         Madrid   NaN      28034    Spain      EMEA          Freyre   
2821       Toulouse   NaN      31000   France      EMEA          Roulet   
2822         Boston    MA      51003      USA       NaN         Yoshido   

     CONTACTFIRSTNAME DEALSIZE CALCULATED_SALES  
0                Kwai    Small          2871.00  
1                Paul    Small          2765.90  
2              Daniel   Medium          3884.34  
3               Julie   Medium          3746.70  
4               Julie   Medium          4900.00  
...               ...      ...              ...  
2818            Diego    Small          2000.00  
2819           Pirkko   Medium          2900.00  
2820            Diego   Medium          4300.00  
2821          Annette    Small          2116.16  
2822             Juri   Medium          3079.44  

[2823 rows x 26 columns]